File: solutions/cpp/Problem69.cpp
#pragma once
/*
69. Sqrt(x)

Implement int sqrt(int x).

Compute and return the square root of x, where x is guaranteed to be a non-negative integer.

Since the return type is an integer, the decimal digits are truncated and only the integer part of the result is returned.
*/
#include <iostream>


class Solution {
public:
	int mySqrt(int x) {
		long long result = x / 2;
		long long inc = x / 4;
		long long tmp = 0;
		while (inc > 1)
		{
			tmp = result * result;
			if (tmp == x)
			{
				return result;
			}
			else
			{
				if (tmp > x)
				{
					result -= inc;
				}
				else
				{
					result += inc;
				}
				inc /= 2;
			}
		}

		if (result * result >= x)
		{
			while (result * result > x)
			{
				result -= 1;
			}
			return result;
		}

		if (result * result < x)
		{
			while (result * result <= x)
			{
				result += 1;
			}
			return result - 1;
		}
	}
	void test()	{
		int i = 2147395599;
		std::cout << i << " sqrt --> " << mySqrt(i) << std::endl;
	}
}; 

File: solutions/python3/problem69.py
# -*- coding: utf-8 -*-
"""
69. Sqrt(x)
Implement int sqrt(int x).

Compute and return the square root of x, where x is guaranteed to be a non-negative integer.

Since the return type is an integer, the decimal digits are truncated and only the integer part of the result is returned.
"""


class Solution:
    def mySqrt(self, x: int) -> int:
        l = 0
        r = x
        res = -1
        while l <= r:
            mid = (l + r) // 2
            if mid ** 2 <= x:
                res = mid
                l = mid + 1
            else:
                r = mid - 1

        return res
